### PR TITLE
feat(stdlib): adds events.stateChanges function

### LINKED ISSUE #3312: Add generic stateChanges function
The function would filter out records that are in the same state as the previous record.

Example test case of the expected behavior

```
import "experimental/array"

data = array.from(
    rows: [
        {_time: 0, state: "x", _value: 10},
        {_time: 1, state: "x", _value: 10},
        {_time: 2, state: "x", _value: 11},
        {_time: 3, state: "y", _value: 12},
        {_time: 4, state: "y", _value: 13},
        {_time: 4, state: "y", _value: 13},
        {_time: 5, state: "z", _value: 14},
        {_time: 6, state: "x", _value: 15},
        {_time: 7, state: "z", _value: 16},
    ],
)
want = array.from(
    rows: [
        {_time: 0, state: "x", _value: 10},
        {_time: 2, state: "y", _value: 12},
        {_time: 4, state: "z", _value: 14},
        {_time: 5, state: "x", _value: 15},
        {_time: 6, state: "z", _value: 16},
    ],
)

test stateChanges = () => ({
    input: data,
    want: want,
    fn: stateChanges,
})
```

### PR DESCRIPTION
**Pull request by nathanielc**
The stateChanges function filters sequential records that have the same
state as the previous record.

Closes #3312

### Done checklist
- [ ] docs/SPEC.md updated
- [x] Test cases written

### CODE REVIEW

File: stdlib/events/state_changes_test.flux [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    rows: [
+        {_time: 2020-10-01T00:00:00Z, _measurement: "foo", _field: "temp", state: "x", _value: 10},
+        {_time: 2020-10-01T00:00:03Z, _measurement: "foo", _field: "temp", state: "y", _value: 12},
+        {_time: 2020-10-01T00:00:05Z, _measurement: "foo", _field: "temp", state: "z", _value: 14},
+        {_time: 2020-10-01T00:00:06Z, _measurement: "foo", _field: "temp", state: "x", _value: 15},
+        {_time: 2020-10-01T00:00:07Z, _measurement: "foo", _field: "temp", state: "z", _value: 16},
+    ],
+)
+
+test stateChanges = () => ({
```
Comment (nathanielc) [inline stdlib/events/state_changes_test.flux:31]: This PR is not quite ready to merge but I wanted to share as I made a few small tweaks to the testing package to make writing tests with `array.from` possible.

File: stdlib/testing/testing.flux [hunk 2]
```diff
@@ -1,39 +1,44 @@
...
+    |> map(
+        fn: (r) => ({r with
+            _field: if exists r._field then r._field else die(msg: "test input table does not have _field column"),
+            _measurement: if exists r._measurement then r._measurement else die(msg: "test input table does not have _measurement column"),
+            _time: if exists r._time then r._time else die(msg: "test input table does not have _time column"),
+        }),
+    )
 
+option load = (tables=<-) => tables |> validate()
+option loadStorage = (csv) => c.from(csv: csv) |> validate()
```
Comment (rockstar) [inline stdlib/testing/testing.flux:21]: Oh, this. This is great. Yeah, this is exactly the right thing to do. I like this so so much.

**Review by rockstar — COMMENTED** [review]
I see this PR is still in draft, but the GitHub bot keeps bugging me about it, so you get a review, darn it! 😄 Just a few small comments.

File: stdlib/events/state_changes_test.flux [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+)
+want = array.from(
+    rows: [
+        {_time: 2020-10-01T00:00:00Z, _measurement: "foo", _field: "temp", state: "x", _value: 10},
+        {_time: 2020-10-01T00:00:03Z, _measurement: "foo", _field: "temp", state: "y", _value: 12},
+        {_time: 2020-10-01T00:00:05Z, _measurement: "foo", _field: "temp", state: "z", _value: 14},
+        {_time: 2020-10-01T00:00:06Z, _measurement: "foo", _field: "temp", state: "x", _value: 15},
+        {_time: 2020-10-01T00:00:07Z, _measurement: "foo", _field: "temp", state: "z", _value: 16},
+    ],
+)
```
Comment (fchikwekwe) [inline stdlib/events/state_changes_test.flux:29]: `array.from()` is so awesome. This is so much easier to read and write than CSV.

File: stdlib/events/stateChanges.go [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+	stateIdx := execute.ColIdx(stateColumn, cols)
+	if stateIdx < 0 {
+		return errors.Newf(codes.FailedPrecondition, "column %q does not exist", stateColumn)
+	}
+
+	colMap := execute.ColMap([]int{0}, builder, tbl.Cols())
+
+	var prev values.Value
+	return tbl.Do(func(cr flux.ColReader) error {
```
[the remark this answers is not in the document]
Comment (fchikwekwe) [reply stdlib/events/stateChanges.go:128]: I'm not a huge fan of the boilerplate for builtins. Seems like a good target for simplification at some point in the future maybe.

File: stdlib/testing/testing.flux (hunk 2 above)
Comment (fchikwekwe) [reply stdlib/testing/testing.flux:21]: Yeah this reads a whole lot better. I'm a fan.

**Comment by tomhollingworth** [comment]
Are there any plans to add in a function parameter for the columns to check the previous value on? 

For example:

```diff
package events_test

import "testing"
import "experimental/array"
import "events"

data = array.from(
    rows: [
        {_time: 2020-10-01T00:00:00Z, _measurement: "foo", _field: "temp", state: "x", _value: 10},
        {_time: 2020-10-01T00:00:01Z, _measurement: "foo", _field: "temp", state: "x", _value: 10},
        {_time: 2020-10-01T00:00:02Z, _measurement: "foo", _field: "temp", state: "x", _value: 11},
        {_time: 2020-10-01T00:00:03Z, _measurement: "foo", _field: "temp", state: "y", _value: 12},
        {_time: 2020-10-01T00:00:04Z, _measurement: "foo", _field: "temp", state: "y", _value: 13},
        {_time: 2020-10-01T00:00:04Z, _measurement: "foo", _field: "temp", state: "x", _value: 13},
        {_time: 2020-10-01T00:00:05Z, _measurement: "foo", _field: "temp", state: "z", _value: 14},
        {_time: 2020-10-01T00:00:06Z, _measurement: "foo", _field: "temp", state: "x", _value: 15},
        {_time: 2020-10-01T00:00:07Z, _measurement: "foo", _field: "temp", state: "z", _value: 16},
    ],
)
want = array.from(
    rows: [
        {_time: 2020-10-01T00:00:00Z, _measurement: "foo", _field: "temp", state: "x", _value: 10},
        {_time: 2020-10-01T00:00:02Z, _measurement: "foo", _field: "temp", state: "x", _value: 11},
        {_time: 2020-10-01T00:00:03Z, _measurement: "foo", _field: "temp", state: "y", _value: 12},
        {_time: 2020-10-01T00:00:04Z, _measurement: "foo", _field: "temp", state: "y", _value: 13},
        {_time: 2020-10-01T00:00:04Z, _measurement: "foo", _field: "temp", state: "x", _value: 13},
        {_time: 2020-10-01T00:00:05Z, _measurement: "foo", _field: "temp", state: "z", _value: 14},
        {_time: 2020-10-01T00:00:06Z, _measurement: "foo", _field: "temp", state: "x", _value: 15},
        {_time: 2020-10-01T00:00:07Z, _measurement: "foo", _field: "temp", state: "z", _value: 16},
    ],
)

test stateChanges = () => ({
  input: data |> testing.load(),
    want: want,
    fn: (tables=<-) => tables
---        |> events.stateChanges()    
+++        |> events.stateChanges(columns: ["_value", "state"])
        |> drop(columns: ["_start", "_stop"]),
})

**Pull request closed without merging**